scripts/pylib/twister/twisterlib/runner.py

```diff
@@ -657,23 +661,34 @@ def process(self, pipeline, done, message, lock, results):
 
         # Run the generated binary using one of the supported handlers
         elif op == "run":
-            logger.debug("run test: %s" % self.instance.name)
-            self.run()
-            logger.debug(f"run status: {self.instance.name} {self.instance.status}")
-            try:
-                # to make it work with pickle
-                self.instance.handler.thread = None
-                self.instance.handler.duts = None
+            if self.options.dyn_filter and not self.dependencies_have_changed():
```

Comment:
Why are you doing this when running? If the purpse of this is static analysis you will be running in --build-only anyways, so you want to do the filtering very early on and not after you have built it, otherwise the cost will be the same.

Reply:
This method is not only intended for static code analysis, and aims to reduce the runtime of long lasting test runs. (For example when running tests on a simulated board.)

If there is a way of getting the ninja deps without actually building, we could and would filter out tests earlier.